// appdata/lib/Service.js
module.exports = (function() {
    'use strict';

    const util = require('util');
    const Types = require('./Type');
    const EventEmitter = require('events').EventEmitter;

    /**
     * Base class for creating new services
     * @constructor Service
     * @augments EventEmitter
     * @param {Object?} options - General service options
     */
    function Service(options) {
        this.type = Types.Service;
        this.options = options || {};
        this.attached = false;
        this.interface = null;
    }

    util.inherits(Service, EventEmitter);
    Service.prototype.constructor = Service;

    /**
     * Attach an interface to the current service
     * @fires attached
     * @memberof Service
     * @param  {Interface} storage - Parent interface
     * @return {this}
     */
    Service.prototype.attach = function(storage) {
        this.interface = storage;
        this.attached = true;
        this.emit('attached', this.interface);
        return this;
    };

    return Service;
})();
